WebActions: Implement storage and REST API

lukasgraf opened the pull request:
Adds an implementation for a **WebActions storage and a REST API** according to [OGIP 42](https://gever.4teamwork.ch/ordnungssystem/3/2/2/5/dossier-2622/document-23827).

**Review Notes**

- A rendered **[preview of the documentation](https://docs.onegovgever.ch/webactions-docs-preview/dev-manual/api/webactions/)** has been published. I recommend starting with this to get a high-level overview of what the idea is here.

- There are certain deviations from what exactly is specified in OGIP 42. I will update it accordingly once the details have been settled. Some of the obvious ones are:
  - Added `enabled` field
  - Split `icon` field into `icon_name` and `icon_data`
  - Renamed the display location `icon` to `title-buttons`
  - Added a server-generated `modified` timestamp (in addition to `created`)
  - Did not implement context-specific actions yet
  - Did not implement `modal` as an allowed scope yet
 
- @jone as discussed, I don't think we need a full review from you - having a look at the published docs for the REST API and making sure that works for you should be enough. And maybe a quick glance at the storage interface at `opengever.webactions.interfaces.IWebActionsStorage`. 

Closes #5214 (Storage)
Closes #5215 (REST API)
Closes #5218 (Permission and Role)
Closes #5217 (Docs)

jone commented on docs/public/dev-manual/api/webactions.rst:
```diff
@@ -0,0 +1,475 @@
+
+.. http:delete:: /@webactions/(action_id)
+
+   Löscht die durch die``action_id`` identifizierte Webaktion.
```
```diff
- die``action_id``
+ die ``action_id``
```

jone commented on docs/public/dev-manual/api/webactions.rst:
```diff
@@ -0,0 +1,475 @@
++=================+===============================+=============================================================================+
+| ``title``       | String, obligatorisch         | Titel der Webaktion                                                         |
++-----------------+-------------------------------+-----------------------------------------------------------------------------+
+| ``unique_name`` | String, optional              | Eindeutiger, vom Ersteller der Webaktion kontrollierter Name                |
```
Was passiert, wenn ich zweimal den gleichen `unique_name` in einem `POST` schicke?

lukasgraf replied:
Zur Zeit ein `500 Internal Server` error :-/ Wird also durch den Storage verhindert, aber noch nicht sehr gut behandelt. Werde dies noch anpassen und entsprechend dokumentieren.

lukasgraf replied:
Updated: Habe das Verhalten angepasst dass ein `400 Bad Request` (mit entsprechender aussagekräftiger Meldung) geschickt wird, und die [Doku entsprechend ergänzt](https://github.com/4teamwork/opengever.core/pull/5298/commits/d2b3dc67a8b8b25bc40544033a9f1cd8a4ea5a2b)

jone commented on docs/public/dev-manual/api/webactions.rst:
```diff
@@ -0,0 +1,475 @@
+| Wert          | Beschreibung                                                     |
++===============+==================================================================+
+| ``self``      | Das Ziel wird direkt im Tab von GEVER geöffnet. Sinnvoll für ein |
+|               | Redirect-Szenario bei demder Benutzer am Schluss wieder          |
```
demder -> dem der

jone commented on docs/public/dev-manual/api/webactions.rst:
```diff
@@ -0,0 +1,475 @@
+|               | Redirect-Szenario bei demder Benutzer am Schluss wieder          |
+|               | zurückgeleitet wird.                                             |
++---------------+------------------------------------------------------------------+
+| ``blank``     | Das Ziel wird in einem neuen Tab geöffnet                        |
```
Ev. ein Punkt am Schluss? Das ist jeweils nicht ganz konsistent 😉

jone commented on docs/public/dev-manual/api/webactions.rst:
```diff
@@ -0,0 +1,475 @@
++---------------+---------------------------------------------------------------------+
+| Wert          | Beschreibung                                                        |
++===============+=====================================================================+
+| ``global``    | Die Webaktionwird grundsätzlich bei allen Objekten angeboten.       |
```
Webaktionwird -> space

jone commented on docs/public/dev-manual/api/webactions.rst:
```diff
@@ -0,0 +1,475 @@
++===============+=====================================================================+
+| ``global``    | Die Webaktionwird grundsätzlich bei allen Objekten angeboten.       |
++---------------+---------------------------------------------------------------------+
+| ``context``   | Noch nicht implementiert. Die Webaktionwird nur angeboten, wenn     |
```
Webaktionwird

jone commented on docs/public/dev-manual/api/webactions.rst:
```diff
@@ -0,0 +1,475 @@
+| ``context``   | Noch nicht implementiert. Die Webaktionwird nur angeboten, wenn     |
+|               | sich der Benutzer auf dem in `context` angegebenen Objekt befindet. |
++---------------+---------------------------------------------------------------------+
+| ``recursive`` | Noch nicht implementiert. Das Ziel wird in einem Modal geöffnet.    |
```
✂️ `Das Ziel wird in einem Modal geöffnet.`

jone commented on docs/public/dev-manual/api/webactions.rst:
```diff
@@ -0,0 +1,475 @@
++---------------+-------------+------------------------------------------------------------------+
+| Feld          | Typ         | Beschreibung                                                     |
++===============+=============+==================================================================+
+| ``action_id`` | Integer     | Pro Mandant eindeutige Identifikation der registrieren Webaktion |
```
registrieren -> registrierten

njohner commented on opengever/webactions/schema.py:
```diff
@@ -0,0 +1,196 @@
+
+        icon_required = ('title-buttons', )
+        icon_forbidden = ('actions-menu', 'add-menu', 'user-menu')
+        icon_optional = ('action-buttons', )  # noqa
```
Instead of `noqa` we could just `return` when we are in a case of `icon_optional`. But up to you

njohner commented on opengever/api/webactions.py:
```diff
@@ -0,0 +1,243 @@
+
+        scrub_json_payload(action_delta)
+
+        # Validate on a copy
```
Why don't we delegate the validation to the storage? Simply doing
```
storage = get_storage()
storage.update(action['action_id'], action_delta)
```
would do the validation and raise exactly as you do here no?

lukasgraf replied:
As discussed: No, the REST API methods don't `raise`, they `return`.

The API semantics are slightly different: A Python API (storage) is supposed to raise Exceptions that can be handled by the caller, but the HTTP REST API needs to return a response with an appropriate status code (not just raise an Exception, which would result in a `500 Internal Server Error`).

The `raise BadRequest()` is a bit special, because this will be caught and turned into an approprate response with status `400`.

So in order to return helpful responses with errors, we do need to do some of the validation in both places to some degree. They also serve slightly different purposes: Validation in the storage layer is the storage protecting itself from bad data. Validation on the REST API layer is more about anticipating what would fail to make it into the storage anyways, and then returning a helpful response to the user, instead of just letting it go bang on the storage layer.

jone replied:
I didn't read the this in detail, but I just read
> As discussed: No, the REST API methods don't raise, they return.

Raised exceptions cause the transaction to be aborted. When returning an error result, transaction handling needs to be done manually. Is this handled correctly?

njohner commented on opengever/api/tests/test_webactions.py:
```diff
@@ -0,0 +1,383 @@
+        }, browser.json)
+
+    @browsing
+    def test_get_webactions(self, browser):
```
`test_get_webaction` singular?

lukasgraf replied:
I agree in theory, but this would be a pain to rebase in at this point 😉

njohner commented on opengever/api/tests/test_webactions.py:
```diff
@@ -0,0 +1,383 @@
+
+        with browser.expect_http_error(code=400, reason='Bad Request'):
+            browser.open(url, method='PATCH', headers=self.HEADERS)
+
```
We'd need a test that we can't f** everything up by patching the `action_id`

lukasgraf commented:
> What I'm missing in the docs is more general informations, such as that GEVER appends GET-parameters to 3rd-party-urls in order to share infos. I assume that will follow in further PRs, is that right?

Yep, that's the plan. I had to cut the documentation short at some point, where some stuff that should go in later is definitely missing at this point. Since passing any parameters to the URL simply isn't implemented yet (this PR just contains storage and REST API), it's kind of wrong to already document something that isn't even implemented yet.

On the other hand, I had to do that for some cases because otherwise the docs just don't make any sense. So yes, I'll definitely expand them as the implementation moves forward.

njohner commented on opengever/api/tests/test_webactions.py:
```diff
@@ -0,0 +1,529 @@
+        url = '%s/@webactions/%s' % (self.portal.absolute_url(), action['action_id'])
+
+        with browser.expect_unauthorized():
+            browser.open(url, method='DELETE', headers=self.HEADERS)
```
We could add a checks that the action was not deleted when unauthorized
```
storage = get_storage()
self.assertItemsEqual([action], storage.list())
```

lukasgraf replied:
That is somewhat implied. Anything else would imply that somehow authentication would have been performed _after_ performing an action. That is basically impossible to get wrong, and actually mostly the responsibility of `plone.rest` (which implements the `permission` attribute  on the `<plone:service />` directive.

njohner commented on opengever/api/tests/test_webactions.py:
```diff
@@ -0,0 +1,529 @@
+
+        with browser.expect_unauthorized():
+            browser.open(url, method='DELETE', headers=self.HEADERS)
+
```
same here

njohner commented on opengever/api/tests/test_webactions.py:
```diff
@@ -0,0 +1,529 @@
+
+        browser.open(url, method='DELETE', headers=self.HEADERS)
+        self.assertEqual(204, browser.status_code)
+
```
here check that it was deleted

lukasgraf commented on opengever/webactions/storage.py:
```diff
@@ -0,0 +1,105 @@
+        # the server generated metadata.
+        new_action['action_id'] = action_id
+        new_action['created'] = now
+        new_action['modified'] = now
```
As discussed: Valid point, not sure 😉 

After thinking about this some more, I don't think we should attempt to make things look the same that aren't. The Dublin Core metadata, it's implementation and our own `changed` field are very separate from this "public"-facing REST API we're dealing here. I don't think there's any benefit to homogenize them.

lukasgraf commented on opengever/webactions/storage.py:
```diff
@@ -0,0 +1,105 @@
+        new_action['action_id'] = action_id
+        new_action['created'] = now
+        new_action['modified'] = now
+        new_action['owner'] = userid if userid else 'Anonymous'
```
At this point, `owner` simply serves both roles. The one who created the action will automatically be assigned as `owner`, and that in turn is then used for access control purposes. I think this works well enough for now, I would only change it if we have a specific need.

lukasgraf commented on opengever/webactions/tests/test_webactions_storage.py:
```diff
@@ -0,0 +1,181 @@
+
+        # Multiple initializations shouldn't remove existing data
+        storage.initialize_storage()
+
```
Yes, but initializing the storage is a current side effect of `__init__`. If we were to change that at some point, that test wouldn't catch that.

lukasgraf commented on opengever/webactions/storage.py:
```diff
@@ -0,0 +1,105 @@
+
+    def list(self):
+        if not self._actions:
+            return []
```
Good point. This also reminded me that I wanted to create an upgrade step to initialize the storage (because we don't want lazy initialization, where writes happen on read).

-> Added an upgrade step to ensure storage is always initialized, dropped the `if not self._actions: return []` part

lukasgraf commented on opengever/webactions/storage.py:
```diff
@@ -0,0 +1,167 @@
+        # Validate schema of the final resulting action on a copy
+        action_copy = action.copy()
+        action_copy.update(action_data)
+        validate_schema(action_copy, IPersistedWebActionSchema)
```
Good point, that was indeed missing. I now fixed the behavior so the `unique_name` constraint always gets checked for both `add` and `update`, and added additionak tests for those cases.

lukasgraf commented on opengever/webactions/storage.py:
```diff
@@ -117,9 +121,14 @@ def list(self):
     def update(self, action_id, action_data):
-        # TODO: Schema Validation
         action = self.get(action_id)
 
```
Nice catch. Added additional validation for this, and some more tests to make sure non-user-controlled fields can't be set or updated.

lukasgraf commented:
@njohner updated - I either implemented or addressed all your comments.

Some trivialities I already rebased and squashed into the existing commits, the more substantial stuff I added as separate commits for easier review. Will squash again at the end before merge.

njohner commented:
Ok looks good to me! Nice work!

The upgrade step is not really necessary I think (There are no classmethods on the Storage, so it will get initialized the first time we do anything with it, as the `__init__` is pretty defensive). Anyway, I tested it locally without the upgrade and the API worked fine. But I guess in that way we are sure it's initialized on all installations.

njohner commented:
Oh and 👍 for making sure the storage returns only copies of the data. I had actually thought about that and decided it was OK as it was (through the api there is no danger, and through the code you have to know what you are doing :-). That was also one reason for asking to not return `[]` on the `list` method, because in one case you would get the persisted list and in the other an empty list in which appending would do nothing...)
